### PR TITLE
[4726 ] Improve form dialogs usability

### PR DESCRIPTION
**Pull request by aachavarria**
https://github.com/craftercms/craftercms/issues/4726

### CODE REVIEW

File: templates/web/legacy/embedded_legacy_form.ftl [hunk 1]
```diff
@@ -158,6 +158,9 @@
                                     });
                                   }
                                 },
+                                pendingChanges: () => {
+                                  window.top.postMessage({ type: 'EMBEDDED_LEGACY_FORM_PENDING_CHANGES', action: 'pendingChanges'}, '*');
```
Comment (rart) [inline templates/web/legacy/embedded_legacy_form.ftl:163]: Did you mean `payload`? What's `action`?
```suggestion
                                  window.top.postMessage({ type: 'EMBEDDED_LEGACY_FORM_PENDING_CHANGES', payload: 'pendingChanges'}, '*');
```

File: templates/web/legacy/embedded_legacy_form.ftl [hunk 2]
```diff
@@ -220,6 +223,9 @@
                     renderComplete: () => {
                       window.top.postMessage({ type: 'EMBEDDED_LEGACY_FORM_RENDERED' }, '*');
                     },
+                    pendingChanges: () => {
+                      window.top.postMessage({ type: 'EMBEDDED_LEGACY_FORM_PENDING_CHANGES', action: 'pendingChanges'}, '*');
```
Comment (rart) [inline templates/web/legacy/embedded_legacy_form.ftl:228]: Same question as above.
```suggestion
                      window.top.postMessage({ type: 'EMBEDDED_LEGACY_FORM_PENDING_CHANGES', payload: 'pendingChanges'}, '*');
```

File: templates/web/legacy/embedded_legacy_form.ftl [hunk 3]
```diff
@@ -248,6 +254,12 @@
               action: 'cancelled'
             }, '*');
           },
+          pendingChanges: function() {
+            window.top.postMessage({
+              type: 'LEGACY_CODE_EDITOR_PENDING_CHANGES',
+              action: 'pendingChanges'
```
Comment (rart) [inline templates/web/legacy/embedded_legacy_form.ftl:261]: As above.
```suggestion
              payload: 'pendingChanges'
```
These don't respect the type/payload pattern. Also, payload should be an object.

File: ui/app/src/components/Dialogs/LegacyCodeEditorDialog.tsx [hunk 4]
```diff
@@ -45,6 +52,10 @@ const translations = defineMessages({
   loadingForm: {
     id: 'craftercms.codeEditor.loadingForm',
     defaultMessage: 'Loading...'
+  },
+  pendingChanges: {
+    id: 'craftercms.codeEditor.pendingChangesConfirmation',
+    defaultMessage: 'You have unsaved changes, do you want to close?'
```
Comment (rart) [inline ui/app/src/components/Dialogs/LegacyCodeEditorDialog.tsx]:
```suggestion
    defaultMessage: 'Close without saving changes?'
```

File: ui/app/src/components/Dialogs/LegacyCodeEditorDialog.tsx [hunk 5]
```diff
@@ -192,17 +209,32 @@ export default function LegacyCodeEditorDialog(props: LegacyCodeEditorDialogProp
     dispatch(minimizeDialog({ id }));
   };
 
+  const onClose = () => {
+    if (pendingChanges) {
+      dispatch(
+        showConfirmDialog({
+          title: formatMessage(translations.pendingChanges),
+          onOk: batchActions([closeConfirmDialog(), closeCodeEditorDialog()]),
```
Comment (rart) [inline ui/app/src/components/Dialogs/LegacyCodeEditorDialog.tsx]: We should also be unlocking the content. Ideally, you would ask the form to close so it uses the existing mechanics to clean up (unlock) and report it got closed like when you press cancel button.

File: ui/app/src/components/Dialogs/LegacyFormDialog.tsx [hunk 6]
```diff
@@ -56,6 +58,10 @@ const translations = defineMessages({
   error: {
     id: 'legacyFormDialog.errorLoadingForm',
     defaultMessage: 'An error occurred trying to load the form'
+  },
+  pendingChanges: {
+    id: 'legacyFormDialog.pendingChangesConfirmation',
+    defaultMessage: 'You have unsaved changes, do you want to close?'
```
Comment (rart) [inline ui/app/src/components/Dialogs/LegacyFormDialog.tsx]:
```suggestion
    defaultMessage: 'Close without saving changes?'
```

File: ui/app/src/components/Dialogs/LegacyFormDialog.tsx [hunk 7]
```diff
@@ -282,17 +296,32 @@ export default function LegacyFormDialog(props: LegacyFormDialogProps) {
     dispatch(minimizeDialog({ id }));
   };
 
+  const onClose = () => {
+    if (pendingChanges) {
+      dispatch(
+        showConfirmDialog({
+          title: formatMessage(translations.pendingChanges),
+          onOk: batchActions([closeConfirmDialog(), closeEditDialog()]),
```
Comment (rart) [inline ui/app/src/components/Dialogs/LegacyFormDialog.tsx]: As above.. Unlock, try using same mechanics as when pressing cancel button.

**Review by rart — CHANGES_REQUESTED** [review]

File: templates/web/legacy/embedded_legacy_form.ftl (hunk 1 above)
Comment (aachavarria) [reply templates/web/legacy/embedded_legacy_form.ftl:163]: this is used as an ID for second type(), for example when the type called is `EMBEDDED_LEGACY_FORM_SAVE`, there are 3 second types:  'save',  'saveAndClose',  'saveAndPreview', and that is the value of action, for the other actions is not needed to send a type, but can be useful

Comment (aachavarria) [reply templates/web/legacy/embedded_legacy_form.ftl:163]: `                      window.top.postMessage({
                        ...response,
                        type: 'EMBEDDED_LEGACY_FORM_SAVE',
                        refresh: false,
                        tab: type,
                        redirectUrl: response.item?.browserUri,
                        action,
                        isNew: true
                      }, '*');`

Comment (aachavarria) [reply templates/web/legacy/embedded_legacy_form.ftl:163]: these changes include moving all the parameters to payload, should I proceed?

**Review by rart — APPROVED** [review]

**Pull request merged**